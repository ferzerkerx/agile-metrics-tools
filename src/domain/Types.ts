import { MetricItem } from './metrics';

export interface TeamMetricsRequest {
  since?: Date;
  until?: Date;
  shouldUpdateEntries: boolean;
  config: any;
}

export interface ElasticSearchService {
  push(payload: MetricItem, shouldReplaceEntry?: boolean): Promise<any>;
  pushMetrics<T extends MetricItem>(
    metricItems: T[],
    shouldReplaceEntry?: boolean
  ): Promise<void>;
}
