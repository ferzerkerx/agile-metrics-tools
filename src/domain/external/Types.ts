export interface ExternalConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  inlineData: ExternalData[];
}

export class ExternalData {
  createdAt: Date;
  teamName: string;
  [type: string]: any;
}

export interface ExternalService {
  fetch(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}

export interface ExternalRepository {
  csv(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}
