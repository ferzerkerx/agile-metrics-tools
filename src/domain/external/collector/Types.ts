export class ExternalCollectorConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  metricType: string;
  inlineData: [];
  constructor({
    since,
    until = null,
    type,
    srcType,
    metricType,
    inlineData = null,
  }) {
    this.since = since ? new Date(since) : null;
    this.until = until ? new Date(until) : null;
    this.type = type;
    this.srcType = srcType;
    this.metricType = metricType;
    this.inlineData = inlineData;
  }
}
