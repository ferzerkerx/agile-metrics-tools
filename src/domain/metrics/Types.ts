import { ExternalCollectorConfig } from '../external/collector/Types';

export interface CollectorService {
  fetch(config: ExternalCollectorConfig): Promise<MetricItem[]>;
}

export interface MetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
}
