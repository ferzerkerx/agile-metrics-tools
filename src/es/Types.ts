import { MetricItem } from '../metrics';

export interface ElasticSearchService {
  push(payload: MetricItem): Promise<any>;
}
