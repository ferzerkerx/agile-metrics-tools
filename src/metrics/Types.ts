export interface CollectorService<
  T extends CollectorConfig,
  K extends MetricItem
> {
  fetch(config: T): Promise<K[]>;
  supports(config: any): boolean;
}

export interface MetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
}

export interface CollectorConfig {
  teamName: string;
}

export interface CollectorModuleFactory<
  T extends CollectorConfig,
  K extends MetricItem
> {
  collectorInstance(): CollectorService<T, K>;
  collectorConfiguration(obj: any): T;
}
