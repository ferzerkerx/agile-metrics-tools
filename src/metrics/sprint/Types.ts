export interface Statistics {
  moveBackward: number;
  moveForward: number;
  movedToDev: Date;
}
