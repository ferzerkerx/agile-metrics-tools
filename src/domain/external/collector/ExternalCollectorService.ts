import { CollectorService } from '../../metrics';
import { ExternalCollectorConfig, ExternalMetricItem } from './Types';
import { ExternalMetricConverter } from './ExternalMetricConverter';
import { ExternalConfig, ExternalService } from '../Types';

export class ExternalCollectorService implements CollectorService {
  constructor(private readonly externalService: ExternalService) {}

  async fetch(
    externalCollectorConfig: ExternalCollectorConfig
  ): Promise<ExternalMetricItem[]> {
    const externalConfig: ExternalConfig = { ...externalCollectorConfig };

    const externalData = await this.externalService.fetch(externalConfig);
    return ExternalMetricConverter.toMetricItem(
      externalData,
      externalCollectorConfig
    );
  }
}
