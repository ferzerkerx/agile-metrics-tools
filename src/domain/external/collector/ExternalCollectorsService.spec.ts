import { ExternalCollectorService } from './ExternalCollectorService';
import { ExternalCollectorConfig } from './Types';
import { ExternalConfig, ExternalData, ExternalService } from '../Types';

function testExternalData(): ExternalData {
  return {
    createdAt: new Date('2018-12-06'),
    teamName: 'someTeamName',
    field1: 'someAdditionalData',
    field2: 4,
    field3: new Date('2018-12-07'),
  };
}

describe('ExternalCollectorsService', () => {
  const externalService: ExternalService = {
    fetch: (externalConfig: ExternalConfig): Promise<ExternalData[]> =>
      Promise.resolve([testExternalData()]),
  };

  const externalCollectorService: ExternalCollectorService =
    new ExternalCollectorService(externalService);

  it('should fetch inline metrics', async () => {
    const externalCollectorConfig: ExternalCollectorConfig =
      new ExternalCollectorConfig({
        since: '2018-11-20',
        type: 'someType',
        srcType: 'inline',
        metricType: 'someType',
        inlineData: [testExternalData()],
      });

    const data = await externalCollectorService.fetch(externalCollectorConfig);
    expect(data).toMatchSnapshot();
  });
});
