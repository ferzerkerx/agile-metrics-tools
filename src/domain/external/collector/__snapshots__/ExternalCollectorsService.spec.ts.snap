// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalCollectorsService should fetch inline metrics 1`] = `
Array [
  Object {
    "createdAt": 2018-12-06T00:00:00.000Z,
    "dataType": "EXT-someType",
    "field1": "someAdditionalData",
    "field2": 4,
    "field3": 2018-12-07T00:00:00.000Z,
    "id": "5c5ab5fc70c2518e50dba3bc08accbac86d8303dd39bbb689daa678c7fcd81f1bcf9122f841888ab5213209173518540a548bb25e5d0e76e5dddaf3b30a90137",
    "teamName": "someTeamName",
  },
]
`;
